### PR TITLE
implementing gradient for Sort op for tensor3

### PR DESCRIPTION
**Pull request by noskill**

### CODE REVIEW

File: theano/tensor/sort.py [hunk 1]
```diff
@@ -80,11 +80,43 @@ def grad(self, inputs, output_grads):
                 idx = argsort(*inputs, kind=self.kind, order=self.order)
                 # not working: numpy.where(idx[None, :]==numpy.arange(2)[:, None, None])
                 pass
+        elif a.ndim == 3:
```
Comment (nouiz) [inline theano/tensor/sort.py]: update the string on line 64 to tell about the new supported case.

File: theano/tensor/tests/test_sort.py [hunk 2]
```diff
@@ -82,6 +82,17 @@ def test_grad_none_axis(self):
...
+        utt.verify_grad(lambda x: sort(x, None), [data])
+
+    def test_grad_negative_axis(self):
+        data = np.random.rand(2, 3, 2).astype(theano.config.floatX)
+        utt.verify_grad(lambda x: sort(x, -1), [data])
+        data = np.random.rand(2, 3, 2).astype(theano.config.floatX)
+        utt.verify_grad(lambda x: sort(x, -3), [data])
+        data = np.random.rand(2, 3, 2).astype(theano.config.floatX)
+        utt.verify_grad(lambda x: sort(x, -2), [data])
+        data = np.random.rand(2, 3, 2).astype(theano.config.floatX)
```
Comment (nouiz) [inline theano/tensor/tests/test_sort.py]: also tests for axis 0, 1 and 2. Just in case.

File: theano/tensor/sort.py [hunk 3]
```diff
@@ -80,11 +80,43 @@ def grad(self, inputs, output_grads):
...
+            if isinstance(axis, theano.Constant) and axis.data is not None:
+                indices = self.__get_argsort_indices(a, axis)
+                inp_grad = output_grads[0].reshape(a.shape)[indices[0], indices[1], indices[2]]
```
Comment (nouiz) [inline theano/tensor/sort.py]: the reshape here isn't needed. output_grads[0] should be of the same shape as a.

File: theano/tensor/tests/test_sort.py [hunk 4]
```diff
@@ -82,6 +82,17 @@ def test_grad_none_axis(self):
         utt.verify_grad(lambda x: sort(x, None), [data])
         #utt.verify_grad(lambda x: sort(x, 0), [data])
         #utt.verify_grad(lambda x: sort(x, 1), [data])
+        data = np.random.rand(2, 3, 2).astype(theano.config.floatX)
```
Comment (nouiz) [inline theano/tensor/tests/test_sort.py]: Can you use shape 2,3,4 for all the new tests? This tests more case.

File: theano/tensor/sort.py [hunk 5]
```diff
@@ -80,11 +80,43 @@ def grad(self, inputs, output_grads):
...
+        else:
+            axis_data = a.ndim + axis.data
+        for i in range(a.ndim):
+            if i == axis_data:
+                indices.append(rev_idx)
+            else:
+                index_shape = [1] * a.ndim 
+                index_shape[i] = a.shape[i]
+                indices.append(theano.tensor.arange(a.shape[i]).reshape(index_shape))
+        return indices
```
Comment (nouiz) [inline theano/tensor/sort.py]: All the indexing here is clever. Can you comment it to help other people understand it?

**Comment by nouiz** [comment]
I made some small comments. Thanks for this PR!

**Comment by noskill** [comment]
Ok, working on this.

**Comment by noskill** [comment]
@nouiz I have made some changes(thanks to tests), should be better now.

**Comment by nouiz** [comment]
For me, it look goods. @lamblin do you want to check it? If I don't get new after next Wednesday, I'll merge.

File: theano/tensor/sort.py (hunk 1 above)
Comment (lamblin) [inline theano/tensor/sort.py]: Done.

**Comment by lamblin** [comment]
Looks good to me too.
Thanks!

**Pull request merged**